lightning/backend/local.go

```diff
@@ -0,0 +1,1149 @@
+	if !ok {
+		// recovery mode, we should reopen this engine file
+		meta, err := local.LoadEngineMeta(engineUUID)
+		if err != nil {
+			// if engine meta not exist, just skip
+			if os.IsNotExist(err) {
+				return nil
+			}
+			return err
+		}
+		db, err := local.openEngineDB(engineUUID, true)
+		if err != nil {
+			return err
+		}
+		engineFile := &LocalFile{
+			localFileMeta: meta,
+			Uuid:          engineUUID,
+			db:            db,
+		}
+		local.engines.Store(engineUUID, engineFile)
+		return nil
+	}
+	engineFile := engine.(*LocalFile)
+	err := engineFile.db.Flush()
+	if err != nil {
+		return err
+	}
+	return local.saveEngineMeta(engineFile)
+}
+
+func (local *local) getImportClient(ctx context.Context, peer *metapb.Peer) (sst.ImportSSTClient, error) {
+	local.grpcClis.mu.Lock()
+	defer local.grpcClis.mu.Unlock()
+	var err error
+
+	conn, ok := local.grpcClis.clis[peer.GetStoreId()]
+	if !ok {
+		conn, err = local.getGrpcConnLocked(ctx, peer.GetStoreId())
+		if err != nil {
+			log.L().Error("could not get grpc connect ", zap.Uint64("storeId", peer.GetStoreId()))
+			return nil, err
+		}
+	}
+	return sst.NewImportSSTClient(conn), nil
+}
+
+func (local *local) WriteToTiKV(
+	ctx context.Context,
+	engineFile *LocalFile,
+	region *split.RegionInfo) ([]*sst.SSTMeta, error) {
+	var startKey, endKey []byte
+	if len(region.Region.StartKey) > 0 {
+		_, startKey, _ = codec.DecodeBytes(region.Region.StartKey, []byte{})
+	}
+	if len(region.Region.EndKey) > 0 {
+		_, endKey, _ = codec.DecodeBytes(region.Region.EndKey, []byte{})
+	}
+	opt := &pebble.IterOptions{LowerBound: startKey, UpperBound: endKey}
+	iter := engineFile.db.NewIter(opt)
+	defer iter.Close()
+
+	iter.First()
+	firstKey := codec.EncodeBytes([]byte{}, iter.Key())
+	iter.Last()
+	lastKey := codec.EncodeBytes([]byte{}, iter.Key())
+
+	meta := &sst.SSTMeta{
+		Uuid:        uuid.NewV4().Bytes(),
+		RegionId:    region.Region.GetId(),
+		RegionEpoch: region.Region.GetRegionEpoch(),
+		Range: &sst.Range{
+			Start: firstKey,
+			End:   lastKey,
+		},
+	}
+
+	leaderID := region.Leader.GetId()
+	clients := make([]sst.ImportSST_WriteClient, 0, len(region.Region.GetPeers()))
+	requests := make([]*sst.WriteRequest, 0, len(region.Region.GetPeers()))
+	for _, peer := range region.Region.GetPeers() {
+		cli, err := local.getImportClient(ctx, peer)
+		if err != nil {
+			return nil, err
+		}
+
+		wstream, err := cli.Write(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// Bind uuid for this write request
+		req := &sst.WriteRequest{
+			Chunk: &sst.WriteRequest_Meta{
+				Meta: meta,
+			},
+		}
+		if err = wstream.Send(req); err != nil {
+			return nil, err
+		}
+		clients = append(clients, wstream)
+		requests = append(requests, req)
+	}
+
+	bytesBuf := newBytesBuffer()
+	defer bytesBuf.destroy()
+	pairs := make([]*sst.Pair, 0, local.batchWriteKVPairs)
+	count := 0
+	size := int64(0)
+	totalCount := 0
+	for iter.First(); iter.Valid(); iter.Next() {
```

Comment:
potential issue and opt:

since we disable the compaction of pebble, all the files are in L0 and may overlap, so at the worst case we will scan the whole DB data here, depend on the order the kvs are written. it's poor to scan the whole db data for every region here.